SM-1238: Go Developer Instructions (#861)
bitwarden/sdk: languages/go/INSTRUCTIONS.md

# Instructions

This guide is for developers who want to use the Bitwarden Go SDK module in their own Go projects. Please see the main [README](./README.md) and [example.go](./example/example.go) file for examples.

## Supported Targets

The Bitwarden Go SDK module utilizes FFI calls to the Bitwarden Rust SDK via [cgo](https://pkg.go.dev/cmd/cgo). The module supports the following statically linked targets:

- Linux `x86-64` & `arm64`
- macOS `x86-64` & `arm64`
- Windows `x86-64`

## Linux

### Prerequisites

- [Go](https://go.dev/dl)
- A C toolchain

We recommend the [MUSL toolchain](https://musl.libc.org). You can install this on most debian based systems with:

```shell
sudo apt install musl-tools
```

### Set Go Environment Info

#### Enable cgo

```shell
go env -w CGO_ENABLED=1
```

#### Set the C compiler

```shell
go env -w CC=musl-gcc
```

#### Verify
```shell
go env
```

### Install the Bitwarden Go SDK

#### Adding the Module

```shell
go get github.com/bitwarden/sdk-go
```

#### Build

```shell
go build -ldflags '-linkmode external -extldflags "-static -Wl,-unresolved-symbols=ignore-all"'
```

> ## Mac

Reviewer: ```suggestion
## macOS
```

Author: Fixed in https://github.com/bitwarden/sdk/pull/861/commits/92b64f40c48540efda0a46405368cd5da5866c59

### Prerequisites

- [Go](https://go.dev/dl)
- A C toolchain

[Clang](https://clang.llvm.org/get_started.html) is the default C and C++ toolchain on Mac OS. The easiest way to ensure you have the toolchain is to install the Xcode Command Line tools.

You can install Clang with:

```shell

> xcode-select -–install

Reviewer: ```suggestion
xcode-select --install
```

Some weird formatting thing going on here. 

Copying and pasting the command fails.

Author: Weird... hopefully fixed in https://github.com/bitwarden/sdk/pull/861/commits/92b64f40c48540efda0a46405368cd5da5866c59

# Instructions

This guide is for developers who want to use the Bitwarden Go SDK module in their own Go projects.

> The Bitwarden SDK is written in Rust, and we utilize FFI via [cgo](https://pkg.go.dev/cmd/cgo). We currently support static linking the Bitwarden Go module on Linux (x86-64 & arm64), Mac (x86-64 & arm64), and Windows (x86-64).

Reviewer: ```suggestion
The Bitwarden Go SDK module utilizes FFI calls to the Rust Bitwarden SDK via [cgo](https://pkg.go.dev/cmd/cgo).

The Bitwarden Go SDK module provides the following statically linked targets:

- Linux (x86-64 & arm64)
- macOS (x86-64 & arm64)
- Windows (x86-64)
```

I personally found the subject of this information unclear.
 
I think we are trying to give the reader more information on why cgo is needed?

Author: Yes, I think it's important to mention that we only support static linking and which targets we support, which brings in the question of what platforms we support. I definitely want it to be concise and clear though.

I've made some changes in https://github.com/bitwarden/sdk/pull/861/commits/730bbce055d05eeaa854d0cd405e5c1edde63953, hopefully that is more clear. Let me know what you think.

Please see the main [README](./README.md) and [example](./example/example.go) for examples.

## Linux

1. Make sure you have the following on your system:
    - [Go](https://go.dev/dl)
    - A C/C++ toolchain. We recommend the [MUSL toolchain](https://musl.libc.org). You can install this on most debian based systems with: `sudo apt install musl-tools`
2. Verify cgo is enabled and the `CC` and `CXX` flags are set with: `go env`.
    - You can enable cgo by setting the flag: `go env -w CGO_ENABLED=1`

>     - If `CC` and `CXX` are not set, you can set them with: `go env -w CC=clang CXX=clang++`

Reviewer 2: ```suggestion
    - If `CC` is not set, you can set it with: `go env -w CC=musl-gcc`
```

Most Linux distros don't use Clang by default. We need to use `musl-gcc` for `CC`. I don't see an equivalent `musl-g++`, but leaving `CXX` unset or the default is fine for our usage.

4. Your project should build with: `go build`

## Windows

1. Make sure you have the following on your system:
    - [Go](https://go.dev/dl)
    - GCC via [MSYS2](https://www.msys2.org)
        - GCC is required on Windows according to [this cgo](https://go.dev/wiki/cgo) page. Microsoft recommends using the MinGW-w64 toolchain via [MSYS2](https://www.msys2.org) for up-to-date native builds of GCC [here](https://code.visualstudio.com/docs/cpp/config-mingw#_installing-the-mingww64-toolchain). MSYS2 installation instructions can be found [here](https://www.msys2.org).
        - After following the MSYS2 installation instructions, you should be able to run the following to verify everything is installed:
            - `gcc --version`
            - `g++ --version`
            - `gdb --version`
2. Verify cgo is enabled and `CC` and `CXX` are set with: `go env`.
    - You can enable cgo by setting the flag: `go env -w CGO_ENABLED=1`
    - You can set `CC` and `CXX` with: `go env -w CC=gcc CXX=g++`
3. Make sure you have added the Bitwarden Go SDK with go get: `go get github.com/bitwarden/sdk-go`

> 4. Your project should build with: `go build`

Reviewer: What about breaking the steps for each OS down into a few sections? 

I personally find this easier to read and code blocks are more convenient to copy.

Windows example below:

## Windows

### Prerequisites

- [Go](https://go.dev/dl)
- GCC

Go [documentation](https://go.dev/wiki/cgo) recommends installing the MinGW-w64 gcc compiler.

We recommend following the Visual Studio Code [guide](https://code.visualstudio.com/docs/cpp/config-mingw#_installing-the-mingww64-toolchain) to install the MinGW-w64 toolchain.

### Set Go Environment for cgo

#### Enable cgo

```shell
go env -w CGO_ENABLED=1
```

#### Set C and C++ Compilers

```shell
go env -w CC=gcc CXX=g++
```

Verify settings by running:

```shell
go env
```

### Install the Bitwarden Go SDK

Add the Bitwarden Go SDK to a project with the following command:

```shell
go get github.com/bitwarden/sdk-go
```

### Build

Validate the project builds successfully

```shell
go build
```

Author: Good point, definitely more readable and more concise.

Updated in: https://github.com/bitwarden/sdk/pull/861/commits/df7c5cd8195b8b0f0adc717ecc978277ead5cb78.
